guard.go

```diff
@@ -0,0 +1,147 @@
+package do
+
+import (
+	"sync"
+)
+
+// A Guard is a conditional variable that has be specialized for the
+// GuardedObject. It should only ever be created by using a GuardedObject and
+// should only ever be used when entering a GuardedObject.
+type Guard struct {
+	condition       func() bool
+	conditionSignal chan struct{}
+}
+
+func newGuard(condition func() bool) Guard {
+	guard := Guard{
+		condition:       condition,
+		conditionSignal: make(chan struct{}, 1),
+	}
+	if guard.condition() {
+		guard.conditionSignal <- struct{}{}
+	}
+	return guard
+}
+
+func (guard *Guard) open() {
+	select {
+	case guard.conditionSignal <- struct{}{}:
+	default:
+	}
+}
+
+func (guard *Guard) close() {
+	select {
+	case <-guard.conditionSignal:
+	default:
+	}
+}
+
+func (guard *Guard) wait() {
+	<-guard.conditionSignal
+}
+
+// A GuardedObject is used to provide safe concurrent read and write access to
+// an object that is accessed concurrently. To be protected, an object should
+// call Enter at the beginning of every method that accesses it, and call Exit
+// at the end of every method that accesses it. A GuardedObject can also define
+// conditional Guards that will block at a call to Enter, until a given
+// condition is met.
+type GuardedObject struct {
+	mu     *sync.RWMutex
+	guards []Guard
+}
+
+// NewGuardedObject returns a new GuardedObject with no Guards defined.
+func NewGuardedObject() GuardedObject {
+	return GuardedObject{
+		mu:     new(sync.RWMutex),
+		guards: make([]Guard, 0),
+	}
+}
+
+// Guard this GuardedObject with a certain condition. The conditional lambda
+// function must not modify any non-local variables. The Guard returned can be
+// passed to a call to the Enter function, blocking the call until the
+// condition is met. The conditions will automatically be re-evaluated. A Guard
+// must only be used with the GuardedObject that created it.
+func (object *GuardedObject) Guard(condition func() bool) *Guard {
+	guard := newGuard(condition)
+	object.mu = new(sync.RWMutex)
```

Comment:
Do we need to create a new mutex every time we create a new Guard?
I think we can get rid of this line.

Reply:
Well spotted. Not sure how that snuck in there 😆